### PR TITLE
feat: added --policy flag to pass policy path and implemented kyverno-json engine

### PR DESCRIPTION
**Pull request by Sanskarzz**
Added --policy flag to pass the path to the policy 
`./kyverno-envoy-plugin serve --policy /path/to/policy1.yaml  --policy /path/to/policy2.yaml`

Also used Load function of this "github.com/kyverno/kyverno-json/pkg/policy" package to load the v1alpha1.ValidatingPolicy type of policies
I also tested ext-authz server with istio , here are the logs of the ext-authz server it seems like policy is successfully loaded in  Check function  .

Here is policy.yaml i used to check the incoming request

```yaml
apiVersion: json.kyverno.io/v1alpha1
kind: ValidatingPolicy
metadata:
  name: test-policy
spec:
  rules:
    - name: deny-external-calls
      assert:
        all:
        - message: "The GET method is restricted to the /foo path."
          check:
            request:
                http:
                    method: 'GET'
                    path: '/foo'  
                        
```

Calling the sample application at /foo

```sh
sanskar@sanskar-HP-Laptop-15s-du1xxx:~/kyverno-envoy-plugin$ kubectl run test -it --rm --restart=Never --image=busybox -- wget -q --header="x-ext-authz: allow" --output-document - echo.demo.svc.cluster.local:8080/foo
{
  "path": "/foo",
  "headers": {
    "host": "echo.demo.svc.cluster.local:8080",
    "user-agent": "Wget",
    "x-ext-authz": "allow",
    "x-forwarded-proto": "http",
    "x-request-id": "a260ace6-c612-455b-954b-b7d46c1f3139",
    "x-b3-traceid": "88910ca85982e843e696ba2077efddb3",
    "x-b3-spanid": "e696ba2077efddb3",
    "x-b3-sampled": "0"
  },
  "method": "GET",
  "body": "",
  "fresh": false,
  "hostname": "echo.demo.svc.cluster.local",
  "ip": "::ffff:127.0.0.6",
  "ips": [],
  "protocol": "http",
  "query": {},
  "subdomains": [
    "svc",
    "demo",
    "echo"
  ],
  "xhr": false,
  "os": {
    "hostname": "echo-6847f9f85-zt8nf"
  },
  "connection": {}
}pod "test" deleted
```
Here are the logs of ext-auth server after calling sample application at /foo
```sh
sanskar@sanskar-HP-Laptop-15s-du1xxx:~$ kubectl logs "$(kubectl get pod -l app=ext-authz -n demo -o jsonpath={.items..metadata.name})" -n demo -c ext-authz -f
Starting HTTP server on Port 8000
Starting GRPC server on Port 9000
Request is initialized in kyvernojson engine .
Request passed the policies.
```
Now calling sample application at /bar
```
sanskar@sanskar-HP-Laptop-15s-du1xxx:~/kyverno-envoy-plugin$ kubectl run test -it --rm --restart=Never --image=busybox -- wget -q --header="x-ext-authz: allow" --output-document - echo.demo.svc.cluster.local:8080/bar

wget: server returned error: HTTP/1.1 403 Forbidden
pod "test" deleted
pod default/test terminated (Error)
```
The Request Failed , policy denied the request 
Here are the logs of  ext-auth after both above requests 
```sh
sanskar@sanskar-HP-Laptop-15s-du1xxx:~$ kubectl logs "$(kubectl get pod -l app=ext-authz -n demo -o jsonpath={.items..metadata.name})" -n demo -c ext-authz -f
Starting HTTP server on Port 8000
Starting GRPC server on Port 9000
Request is initialized in kyvernojson engine .
Request passed the policies.
Request is initialized in kyvernojson engine .
Request denied with reason: -> The GET method is restricted to the /foo path.
 -> all[0].check.request.http.path: Invalid value: "/bar": Expected value: "/foo"
```

### CODE REVIEW

**Comment by anushkamittal2001** [comment]
Can you check the linter errors.

**Comment by Sanskarzz** [comment]
> Can you check the linter errors.

I tried to fix lint error first i deleted the go.mod and go.sum files and re-initializing the module with `go mod init` and `go mod tidy`
Below are the logs of `go mod tidy` The lint is failing because of checsum mismatch of kyverno-json package 
```
sanskar@sanskar-HP-Laptop-15s-du1xxx:~/kyverno-envoy-plugin$ go mod tidy
go: finding module for package github.com/envoyproxy/go-control-plane/envoy/service/auth/v3
go: finding module for package github.com/envoyproxy/go-control-plane/envoy/type/v3
go: finding module for package github.com/kyverno/kyverno-json/pkg/policy
go: finding module for package github.com/kyverno/kyverno-json/pkg/json-engine
go: downloading github.com/kyverno/kyverno-json v0.0.2
go: finding module for package github.com/spf13/cobra
go: finding module for package go.uber.org/multierr
go: finding module for package google.golang.org/genproto/googleapis/rpc/status
go: finding module for package google.golang.org/grpc
go: finding module for package google.golang.org/grpc/codes
go: finding module for package k8s.io/apimachinery/pkg/util/wait
go: finding module for package github.com/golang-jwt/jwt
go: finding module for package github.com/jmespath-community/go-jmespath/pkg/functions
go: finding module for package google.golang.org/protobuf/encoding/protojson
go: finding module for package github.com/jmespath-community/go-jmespath/pkg/interpreter
go: finding module for package github.com/jmespath-community/go-jmespath/pkg/parsing
go: finding module for package github.com/kyverno/kyverno-json/pkg/engine/template
go: found github.com/envoyproxy/go-control-plane/envoy/service/auth/v3 in github.com/envoyproxy/go-control-plane v0.12.0
go: found github.com/envoyproxy/go-control-plane/envoy/type/v3 in github.com/envoyproxy/go-control-plane v0.12.0
go: found github.com/spf13/cobra in github.com/spf13/cobra v1.8.0
go: found go.uber.org/multierr in go.uber.org/multierr v1.11.0
go: found google.golang.org/genproto/googleapis/rpc/status in google.golang.org/genproto/googleapis/rpc v0.0.0-20240401170217-c3f982113cda
go: found google.golang.org/grpc in google.golang.org/grpc v1.63.0
go: found google.golang.org/grpc/codes in google.golang.org/grpc v1.63.0
go: found k8s.io/apimachinery/pkg/util/wait in k8s.io/apimachinery v0.29.3
go: found github.com/golang-jwt/jwt in github.com/golang-jwt/jwt v3.2.2+incompatible
go: found github.com/jmespath-community/go-jmespath/pkg/functions in github.com/jmespath-community/go-jmespath v1.1.1
go: found google.golang.org/protobuf/encoding/protojson in google.golang.org/protobuf v1.33.0
go: found github.com/jmespath-community/go-jmespath/pkg/interpreter in github.com/jmespath-community/go-jmespath v1.1.1
go: found github.com/jmespath-community/go-jmespath/pkg/parsing in github.com/jmespath-community/go-jmespath v1.1.1
go: finding module for package github.com/kyverno/kyverno-json/pkg/engine/template
go: finding module for package github.com/kyverno/kyverno-json/pkg/json-engine
go: finding module for package github.com/kyverno/kyverno-json/pkg/policy
go: github.com/kyverno/kyverno-envoy-plugin imports
        github.com/kyverno/kyverno-json/pkg/json-engine: github.com/kyverno/kyverno-json@v0.0.2: verifying module: checksum mismatch
        downloaded: h1:Rcx51owh1Ccjwxtd85crN2jjQ+F6Up1B4oLlNZhX7/4=
        sum.golang.org: h1:r+RU0b4QzK+oZxrW6WZVOqlhhGLdZU3gRLJAMY8NeYE=

SECURITY ERROR
This download does NOT match the one reported by the checksum server.
The bits may have been replaced on the origin server, or an attacker may
have intercepted the download attempt.

For more information, see 'go help module-auth'.
go: github.com/kyverno/kyverno-envoy-plugin imports
        github.com/kyverno/kyverno-json/pkg/policy: github.com/kyverno/kyverno-json@v0.0.2: verifying module: checksum mismatch
        downloaded: h1:Rcx51owh1Ccjwxtd85crN2jjQ+F6Up1B4oLlNZhX7/4=
        sum.golang.org: h1:r+RU0b4QzK+oZxrW6WZVOqlhhGLdZU3gRLJAMY8NeYE=

SECURITY ERROR
This download does NOT match the one reported by the checksum server.
The bits may have been replaced on the origin server, or an attacker may
have intercepted the download attempt.

For more information, see 'go help module-auth'.
go: github.com/kyverno/kyverno-envoy-plugin/pkg/scratch imports
        github.com/kyverno/kyverno-json/pkg/engine/template: github.com/kyverno/kyverno-json@v0.0.2: verifying module: checksum mismatch
        downloaded: h1:Rcx51owh1Ccjwxtd85crN2jjQ+F6Up1B4oLlNZhX7/4=
        sum.golang.org: h1:r+RU0b4QzK+oZxrW6WZVOqlhhGLdZU3gRLJAMY8NeYE=

SECURITY ERROR
This download does NOT match the one reported by the checksum server.
The bits may have been replaced on the origin server, or an attacker may
have intercepted the download attempt.

For more information, see 'go help module-auth'.
```
I also tried after `go clean -modcache` to clear the module cache and re-downloaded modules then also the logs error were same . I am using latest version of kyverno-json which is 0.0.2 , i think this error is because of kyverno-json side .

**Comment by Sanskarzz** [comment]
> Can you check the linter errors.

The CI is green , Lint error is fixed

File: main.go [hunk 1]
```diff
@@ -69,20 +73,96 @@ func handler(w http.ResponseWriter, r *http.Request) {
...
+		return nil, err
 	}
-	// Print the entire struct with field names
-	fmt.Printf("Attributes: %+v\n", attrs)
 
-	// Implement your authorization logic here
-	// For now, allow all requests
+	engine := jsonengine.New()
+	var responses []jsonengine.Response
+	responses = append(responses, engine.Run(ctx, jsonengine.Request{
```
Comment (anushkamittal2001) [inline main.go:94]: won't we have just one response every time check runs? for each request?

Comment (Sanskarzz) [reply main.go]: yes their should be one response , i have made changes

File: main.go [hunk 2]
```diff
@@ -69,20 +73,90 @@ func handler(w http.ResponseWriter, r *http.Request) {
...
+
+	for _, policy := range response.Policies {
+		for _, rule := range policy.Rules {
+			if rule.Error != nil {
+				// If there is an error, deny the request and set the denial reason
+				allowed = false
+				denialReason = rule.Error.Error()
+				fmt.Printf("Request denied by error: %s\n", denialReason)
+				break
+			} else if len(rule.Violations) != 0 {
```
Comment (anushkamittal2001) [inline main.go:111]: Let's manage multiple violations with a single request

Comment (anushkamittal2001) [reply main.go:111]: Apart from that lgtm!

File: main.go [hunk 3]
```diff
@@ -97,43 +97,32 @@ func (s *extAuthzServerV3) Check(ctx context.Context, req *authv3.CheckRequest)
...
-				fmt.Printf("Request denied with reason: %s\n", denialReason)
-				break
+				fmt.Printf("Request violation: %v\n", violations)
 			} else {
-				// If the rule passed, allow the request (unless overridden by a subsequent rule)
-				allowed = true
-				denialReason = ""
-				fmt.Printf("Request passed the policies.\n")
+				// If the rule passed, log it but continue to the next rule/policy
+				fmt.Printf("Request passed the %v policy rule.\n", rule.Rule.Name)
```
Comment (anushkamittal2001) [inline main.go:116]: Let's not print it, instead let's log it

**Review by anushkamittal2001 — APPROVED** [review]

**Review by anushkamittal2001 — DISMISSED** [review]

**Review by anushkamittal2001 — APPROVED** [review]

**Review by anushkamittal2001 — APPROVED** [review]

**Review by anushkamittal2001 — APPROVED** [review]

**Pull request merged**